ext/zend_test/observer.c

```diff
@@ -304,9 +304,12 @@ static ZEND_INI_MH(zend_test_observer_OnUpdateCommaList)
 	zend_array **p = (zend_array **) ZEND_INI_GET_ADDR();
 	zend_string *funcname;
 	zend_function *func;
+	if (!ZT_G(observer_enabled)) {
+		return FAILURE;
+	}
 	if (stage != PHP_INI_STAGE_STARTUP && stage != PHP_INI_STAGE_ACTIVATE && stage != PHP_INI_STAGE_DEACTIVATE && stage != PHP_INI_STAGE_SHUTDOWN) {
 		ZEND_HASH_FOREACH_STR_KEY(*p, funcname) {
-			if ((func = zend_hash_find_ptr(EG(function_table), funcname))) {
+			if ((func = zend_hash_find_ptr(EG(function_table), funcname)) && ZEND_OBSERVER_DATA(func) != NULL) {
```

Comment:
I don't understand this. `ZEND_OBSERVER_DATA()` expands to `&RUN_TIME_CACHE(op_array)[handle]`, which cannot be `NULL` (except for `RUN_TIME_CACHE(op_array) == NULL && handle == 0`, which is undefined behavior). Removing these two conditions still makes the test pass. Should this have dereferenced the pointer?

Reply:
Actually, it probably must handle `ZEND_OBSERVER_NOT_OBSERVED` and `ZEND_OBSERVER_NONE_OBSERVED`. But I forgot about the details of this api.

Reply:
I'll have a look as soon as possible.

Reply:
@iluuu1994, this PR targets PHP-8.4; the definition of `ZEND_OBSERVER_DATA()` has changed there:

https://github.com/php/php-src/blob/58929919410a57446664850bcad6dc516e9ca1a7/Zend/zend_observer.h#L39-L40

That part is needed if `zend_test.observer.enabled=1` upfront (default for our Circle CI job).

@bwoebi, when running the test with `zend_test.observer.enabled=1`, `ZEND_OBSERVER_DATA(func)` is `NULL` for me.

Reply:
It's because `RUN_TIME_CACHE(&func->common)` is `NULL`, so doing pointer arithmetics on it is UB. In that case, I think we should check whether `RUN_TIME_CACHE(&func->common)` is set.

Reply:
Ah, thank you! I'll have a closer look again, and might actually go with the [hash table](https://github.com/php/php-src/pull/16438#issuecomment-2417746841).